datafusion/sqllogictest/test_files/scalar.slt

```diff
@@ -1864,10 +1864,10 @@ query TT
 EXPLAIN SELECT letter, letter = LEFT(letter2, 1) FROM simple_string;
 ----
 logical_plan
-01)Projection: simple_string.letter, simple_string.letter = left(simple_string.letter2, Int64(1))
+01)Projection: simple_string.letter, simple_string.letter = left(CAST(simple_string.letter2 AS Utf8View), Int64(1))
```

Comment:
I'm thinking maybe we should avoid casting type if they have the same Logical type like utf8 -> utf8view. 🤔

Reply:
100% -- casts can often require trivial computation during query -- in this particular case casting `letter2` to a Utf8View means it will copy at least an additional 128 bytes for each row (each view is 128 bytes)

Reply:
Thanks for the suggestion. I implement not to cast if their logical types are the same. However, it is failing in some cases where `Dictionary` is in the signature. In those cases, the logical type is the same, but the native type is `Dictionary` causing type mismatch. The test can be reproduced 
```
cargo test --test sqllogictests -- jctest
```
Any suggestions?

Reply:
We might need to implement the kernel for dictionary type.

Reply:
I see. Can you elaborate on how we `implement the kernel for dictionary type`? I can give it a try.

Reply:
Check the `fn invoke` function for each function.

For example `lpad`
```rust
fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
        match args[0].data_type() {
            Utf8 | Utf8View => make_scalar_function(lpad::<i32>, vec![])(args),
            LargeUtf8 => make_scalar_function(lpad::<i64>, vec![])(args),
            other => exec_err!("Unsupported data type {other:?} for function lpad"),
        }
    }
```

It support utf8/utf8view/largeutf8, but not dictionary. You can rewrite it like this

```rust
fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
        invoke_inner(args, &args[0].data_type())
}

fn invoke_inner(args: &[ColumnarValue], data_type: &DataType) -> Result<ColumnarValue> {
    match data_type {
        DataType::Dictionary(_, v) => {
            invoke_inner(args, v.as_ref())
        }
        Utf8 | Utf8View => make_scalar_function(lpad::<i32>, vec![])(args),
        LargeUtf8 => make_scalar_function(lpad::<i64>, vec![])(args),
        other => exec_err!("Unsupported data type {other:?} for function lpad"),
    }
}
```

Reply:
Thanks for the pointer! I added the support for `left` function. Can you give a read of whether that is a good design?

Reply:
Hey @jayzhan211, I have added dictionary support for those four functions and `repeat` because the logical type casting skipping fails some relevant tests. Please let me know if you have any thoughts on the current implementation. 

I also have a question when getting values from `Dictionary` to get the actual string, it drops `NULL` values, which causes some tests to fail. Is there any helper method I can use to get `NULL` preserving values from the `Dictionary`?

Reply:
https://github.com/apache/datafusion/blob/bfabd48de2e0deccbd8ce595936ad24f18baca65/datafusion/functions-nested/src/string.rs#L394 possibly  @jiashenC ?

Reply:
It seems that iterating values in Dict is quite tricky

```rust
        let test = vec!["a", "a", "b", "c"];
        let array : DictionaryArray<Int32Type> = test.iter().map(|&x| if x == "b" {None} else {Some(x)}).collect();
        let typed = array.downcast_dict::<StringArray>().unwrap();
        for v in typed {
            println!("iter value in dict: {:?}", v);
        }
```